### PR TITLE
Ring buffer

### PR DESCRIPTION
**Pull request by Volkalex28**

### CODE REVIEW

File: Core/Inc/ring_buffer.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  uint8_t fulled : 1;
+  uint8_t empty  : 1;
+  
+} RingBuffer_t;
+
+uint32_t RING_Init(RingBuffer_t* pRing, uint8_t* pMemory, uint16_t size);
+void RING_Clear(RingBuffer_t* pRing);
+void RING_Append(RingBuffer_t* pRing, uint8_t value);
+uint8_t RING_Pop(RingBuffer_t* pRing);
+uint16_t RING_GetCount(RingBuffer_t* pRing);
```
Comment (d-serj) [inline Core/Inc/ring_buffer.h]: I assume that `const` qualifier should be added to a function parameter

File: Core/Inc/ring_buffer.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+#endif //_RING_BUFFER_H_
```
Comment (d-serj) [inline Core/Inc/ring_buffer.h]: Don't forget about empty lines

File: Core/Inc/uart_ring.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+#define UART_RING_SIZE 20
+
+typedef struct
+{
+  STM32F051_UART_t uart;
+  RingBuffer_t ring;
+  uint8_t buff[UART_RING_SIZE];
+
+  void (*UART_RxCallback)();
```
Comment (d-serj) [inline Core/Inc/uart_ring.h]: Minor: add `void` as a function parameter to show that the function doesn't get any input parameter

File: Core/Src/main.c [hunk 4]
```diff
@@ -9,46 +9,57 @@
...
 #include "pins.h"
 
 #include "ds3231.h"
+#include "uart_ring.h"
+
+struct
+{
+  char strOut[100];
+  uint8_t count;
+  char bufIn[100];
```
Comment (d-serj) [inline Core/Src/main.c]: Please, add a description for this magic variable `100` or add a definition/enumeration/constant that will hold this number

File: Core/Src/main.c [hunk 5]
```diff
@@ -9,46 +9,57 @@
...
-    time.Minutes = 9;
-    time.Seconds = 20;
+    DS3231.Time.Hours = 4;
+    DS3231.Time.Minutes = 47;
+    DS3231.Time.Seconds = 20;
 
-    DS3231_SetTime(&DS3231, &time);
+    DS3231_SetTime(&DS3231, &DS3231.Time);
   }
+  else {
```
Comment (d-serj) [inline Core/Src/main.c]: Don't forget about the code style we have chosen.
`{` must be placed at the beginning of the next line

File: Core/Src/main.c [hunk 6]
```diff
@@ -61,34 +72,44 @@ int main()
 
   RCC_Init();
   GPIO_Init();
-  UART1_Init();
   I2C1_Init();
+  UART1_Init(&uart1.uart);
```
Comment (d-serj) [inline Core/Src/main.c]: I would suggest to create a separate program module, that may be called `uart_commander` or something like that. Place these initializations of the ring buffer and uart to that new program module

File: Core/Src/ring_buffer.c [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  pRing->pMemory = pMemory;
+  pRing->size = size;
+  RING_Clear(pRing);
+
+  return 0;
+}
+
+void RING_Append(RingBuffer_t* pRing, uint8_t value)
+{
+  if(pRing == NULL || pRing->fulled == 1)
```
Comment (d-serj) [inline Core/Src/ring_buffer.c]: Add `ASSERT(pRing)` instead of this condition

File: Core/Src/ring_buffer.c [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  if(pRing->tail == pRing->head)
+  {
+    pRing->fulled = 1;
+  }
+  pRing->empty = 0;
+}
+
+uint8_t RING_Pop(RingBuffer_t* pRing)
+{
+  if(pRing == NULL || pRing->empty == 1)
```
Comment (d-serj) [inline Core/Src/ring_buffer.c]: Add `ASSERT(pRing)`

File: Core/Src/ring_buffer.c [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+  if(pRing->empty == 1)
+  {
+    return 0;
+  }
+  else if(pRing->fulled == 1)
+  {
+    return pRing->size;
+  }
+  else
```
Comment (d-serj) [inline Core/Src/ring_buffer.c:180]: There is no need to add `else` here. Remove it

File: Core/Src/uart_ring.c [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+{
+  RING_Append(&((UART_Ring_t*)pUART)->ring, (uint8_t)pUART->handle->RDR);
+  UART_ReceiveIT(pUART, &temp, 1);
+
+  ((UART_Ring_t*)pUART)->UART_RxCallback();
+}
+
+uint32_t UART_Ring_Init(UART_Ring_t* ptr)
+{
+  #ifdef DEBUG
```
Comment (d-serj) [inline Core/Src/uart_ring.c]: Your `ASSERT` definition must contain checking for `DEBUG`. So remove this preprocessor condition and move it to ASSERT module

File: Core/Src/uart_ring.c [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    ASSERT(ptr);
+    ASSERT(ptr->UART_RxCallback);
+  #endif
+
+  if(RING_Init(&ptr->ring, ptr->buff, UART_RING_SIZE) != 0)
```
Comment (d-serj) [inline Core/Src/uart_ring.c]: What could be wrong with `RING_Init`? Why are you checking its return result?

File: Drivers/STM32F051/Src/stm32f051_gpio.c [hunk 12]
```diff
@@ -38,6 +39,9 @@
  */
 inline void GPIO_SetPin(GPIO_TypeDef* GPIOx, uint16_t GPIO_Pin) 
 {
+  #ifdef DEBUG
```
Comment (d-serj) [inline Drivers/STM32F051/Src/stm32f051_gpio.c]: Remove `#ifdef DEBUG` from here

File: Drivers/STM32F051/Src/stm32f051_gpio.c [hunk 13]
```diff
@@ -50,6 +54,9 @@ inline void GPIO_SetPin(GPIO_TypeDef* GPIOx, uint16_t GPIO_Pin)
  */
 inline void GPIO_ResetPin(GPIO_TypeDef* GPIOx, uint16_t GPIO_Pin) 
 {
+  #ifdef DEBUG
```
Comment (d-serj) [inline Drivers/STM32F051/Src/stm32f051_gpio.c]: Remove `#ifdef DEBUG` from here

File: Drivers/STM32F051/Src/stm32f051_gpio.c [hunk 14]
```diff
@@ -62,6 +69,9 @@ inline void GPIO_ResetPin(GPIO_TypeDef* GPIOx, uint16_t GPIO_Pin)
  */
 inline void GPIO_TogglePin(GPIO_TypeDef* GPIOx, uint16_t GPIO_Pin) 
 {
+  #ifdef DEBUG
```
Comment (d-serj) [inline Drivers/STM32F051/Src/stm32f051_gpio.c]: Remove `#ifdef DEBUG` from here

File: Drivers/STM32F051/Src/stm32f051_gpio.c [hunk 15]
```diff
@@ -77,6 +87,9 @@ inline void GPIO_TogglePin(GPIO_TypeDef* GPIOx, uint16_t GPIO_Pin)
  */
 inline GPIO_PinState GPIO_ReadPin(GPIO_TypeDef* GPIOx, uint16_t GPIO_Pin) 
 {
+  #ifdef DEBUG
```
Comment (d-serj) [inline Drivers/STM32F051/Src/stm32f051_gpio.c]: Remove `#ifdef DEBUG` from here

File: Drivers/STM32F051/Src/stm32f051_uart.c [hunk 16]
```diff
@@ -78,34 +44,75 @@ STM32F051_UART_t uart1;
  * @retval 0 If no errors occurred
  * @retval 1 If an error occurs during the process
  */
-uint32_t UART1_WaitFlag(const uint32_t Flag, const FlagStatus Status,
-  const uint32_t tickStart, const uint32_t Timeout)
+uint32_t UART_WaitFlag(const STM32F051_UART_t* pUART, const uint32_t Flag, 
+  const FlagStatus Status, const uint32_t tickStart, const uint32_t Timeout)
 {
-  while((READ_BIT(USART1->ISR, Flag) ? SET : RESET) == Status) 
+  #ifdef DEBUG
```
Comment (d-serj) [inline Drivers/STM32F051/Src/stm32f051_uart.c]: Remove `#ifdef DEBUG` from here

File: Drivers/STM32F051/Src/stm32f051_uart.c [hunk 17]
```diff
@@ -78,34 +44,75 @@ STM32F051_UART_t uart1;
...
+ * 
+ * @param[in] pUART Pointer to UART module object
+ * 
+ * @return UART initialization status
+ * @retval 0 If initialization completed successfully
+ * @retval 1 If an error occurs during the process
+ */
+uint32_t UART_Init(const STM32F051_UART_t* pUART)
+{
+  #ifdef DEBUG
```
Comment (d-serj) [inline Drivers/STM32F051/Src/stm32f051_uart.c]: Remove `#ifdef DEBUG` from here

File: Core/Src/main.c [hunk 18]
```diff
@@ -9,46 +9,57 @@
...
 
 #include "ds3231.h"
+#include "uart_ring.h"
+
+struct
+{
+  char strOut[100];
+  uint8_t count;
+  char bufIn[100];
+} bufUART;
```
Comment (d-serj) [inline Core/Src/main.c:66]: I would suggest creating a separate program module to keep all UART related stuff: ring buffer, uart, and so on.

**Review by d-serj — CHANGES_REQUESTED** [review]
Add a Doxygen comment for all the public functions, structures and its fields

File: Core/Src/main.c (hunk 18 above)
Comment (Volkalex28) [reply Core/Src/main.c:66]: I don't agree. One of the main tasks set for the software is the implementation of UART with a ring buffer. For this reason, I consider it advisable to leave it here. But maybe I'll clean it up when the OS appears

File: Core/Src/uart_ring.c [hunk 19]
```diff
@@ -1,30 +1,129 @@
...
+ * @retval 0 If the buffer is empty
+ * @retval uint8_t Byte value
+ */
+uint8_t UART_Ring_PopByte(UART_Ring_t* const ptr)
+{
+  ASSERT(ptr);
+
+  if(RING_GetCount(&ptr->ring) != 0)
+  {
+    return RING_Pop(&ptr->ring);
```
Comment (d-serj) [inline Core/Src/uart_ring.c]: what if there is `\0` in the ring buffer?
I would add a variable to receive data as a function parameter
`uint8_t UART_Ring_PopByte(UART_Ring_t* const ptr, uint8_t *var_to_receive)`

File: Core/Inc/ring_buffer.h (hunk 2 above)
Comment (d-serj) [reply Core/Inc/ring_buffer.h]: > 
> 
> Don't forget about empty lines

File: Core/Src/ring_buffer.c (hunk 9 above)
Comment (d-serj) [reply Core/Src/ring_buffer.c:180]: > 
> 
> There is no need to add `else` here. Remove it

File: Core/Src/user_assert.c [hunk 20]
```diff
@@ -1,10 +1,80 @@
...
+ * 
+ * This function saves data about the place where the assert 
+ * was fired into a variable @link g_assert_info  @endlink
+ * 
+ * @param[in] pc Programm counter
+ * @param[in] lr Function return address
+ * 
+ * @return None
+ */
 void my_assert(uint32_t *pc, uint32_t *lr) {
```
Comment (d-serj) [inline Core/Src/user_assert.c]: Please, place `{` on the newline

File: Drivers/STM32F051/Src/stm32f051_gpio.c [hunk 21]
```diff
@@ -38,6 +41,7 @@
  */
 inline void GPIO_SetPin(GPIO_TypeDef* GPIOx, uint16_t GPIO_Pin) 
 {
+  ASSERT(GPIOx);
   GPIOx->BSRR = (uint32_t)GPIO_Pin;
```
Comment (d-serj) [inline Drivers/STM32F051/Src/stm32f051_gpio.c]: It is too dangerous to cast smaller types to bigger ones. Don't do that

File: Drivers/STM32F051/Src/stm32f051_gpio.c [hunk 22]
```diff
@@ -50,6 +54,7 @@ inline void GPIO_SetPin(GPIO_TypeDef* GPIOx, uint16_t GPIO_Pin)
  */
 inline void GPIO_ResetPin(GPIO_TypeDef* GPIOx, uint16_t GPIO_Pin) 
 {
+  ASSERT(GPIOx);
   GPIOx->BRR = (uint32_t)GPIO_Pin;
```
Comment (d-serj) [inline Drivers/STM32F051/Src/stm32f051_gpio.c]: Don't cast smaller types to bigger

**Review by d-serj — CHANGES_REQUESTED** [review]

**Pull request merged**